esmvalcore/config/_config_object.py

```diff
@@ -93,29 +112,140 @@ def _load_default_config(cls, filename: Union[os.PathLike, str]):
+        loading the module without any configuration file (this is relevant if
+        the module is used within a script or notebook). To check if the file
+        actually exists, use the method `load_from_file` (this is done when
+        using the `esmvaltool` CLI).
+
+        If used within the esmvaltool program, set the
+        _ESMVALTOOL_USER_CONFIG_FILE_ at the end of this method to make sure
+        that subsequent calls of this method (also in suprocesses) use the
+        correct user configuration file.
+
+        """
+        # (1) Try to get user configuration file from `filename` argument (also
+        # test relative to ~/.esmvaltool)
+        config_user = filename
+
+        # (2) Try to get user configuration file from internal
+        # _ESMVALTOOL_USER_CONFIG_FILE_ environment variable
+        if (
+                config_user is None and
+                '_ESMVALTOOL_USER_CONFIG_FILE_' in os.environ
+        ):
+            config_user = os.environ['_ESMVALTOOL_USER_CONFIG_FILE_']
+
+        # (3) Try to get user configuration file from CLI arguments
+        if config_user is None:
+            config_user = Config._get_config_path_from_cli()
+
+        # (4) Default location
+        if config_user is None:
+            config_user = Config._DEFAULT_USER_CONFIG_DIR / 'config-user.yml'
+
+        config_user = Path(config_user).expanduser()
+
+        # Use path relative to ~/.esmvaltool if necessary
+        if not (config_user.is_file() or config_user.is_absolute()):
+            config_user = Config._DEFAULT_USER_CONFIG_DIR / config_user
+        config_user = config_user.absolute()
+
+        # If used within the esmvaltool program, make sure that subsequent
+        # calls of this method (also in suprocesses) use the correct user
+        # configuration file
+        if Path(sys.argv[0]).name == 'esmvaltool':
+            os.environ['_ESMVALTOOL_USER_CONFIG_FILE_'] = str(config_user)
+
+        return config_user
+
+    @staticmethod
+    def _get_config_path_from_cli() -> None | str:
+        """Try to get configuration path from CLI arguments.
+
+        The hack of directly parsing the CLI arguments here (instead of using
+        the fire or argparser module) ensures that the correct user
+        configuration file is used. This will always work, regardless of when
+        this module has been imported in the code.
+
+        Note
+        ----
+        This only works if the script name is `esmvaltool`. Does not check if
+        file exists.
+
+        """
+        if Path(sys.argv[0]).name != 'esmvaltool':
+            return None
+
+        for arg in sys.argv:
+            for opt in ('--config-file', '--config_file'):
+                if opt in arg:
+                    # Parse '--config-file=/file.yml' or
+                    # '--config_file=/file.yml'
+                    partition = arg.partition('=')
+                    if partition[2]:
+                        return partition[2]
```

Comment:
yeah - see - it could return an empty string if `--config-file` has no value - remember that we can run whatever we want with Fire eg `esmvaltool --config-file` - this will raise a Fire error (not a bool operator etc, line 518 in _main.py):
```
    try:
        fire.Fire(ESMValTool())
    except fire.core.FireExit:
        raise
 ```
 but I think that's way down the stack for this case

Reply:
Very good point! I added some code to simply return `None` now here in https://github.com/ESMValGroup/ESMValCore/pull/2309/commits/2ed5a78d63b792bad27363551cf11da7ff39c74e and let fire handle this further down the line. Before that, a very ugly `IndexError` has been raised.

Reply:
good man!